diff --git a/arch/arm/configs/orion5x_defconfig b/arch/arm/configs/orion5x_defconfig
index 5383cd0..1cbf3f6 100644
--- a/arch/arm/configs/orion5x_defconfig
+++ b/arch/arm/configs/orion5x_defconfig
@@ -192,6 +192,8 @@ CONFIG_MACH_TS209=y
 CONFIG_MACH_TERASTATION_PRO2=y
 CONFIG_MACH_LINKSTATION_PRO=y
 CONFIG_MACH_LINKSTATION_MINI=y
+CONFIG_MACH_LINKSTATION_PRODUO=y
+CONFIG_MACH_LINKSTATION_PRODUO_REV1=y
 CONFIG_MACH_TS409=y
 CONFIG_MACH_WRT350N_V2=y
 CONFIG_MACH_TS78XX=y
diff --git a/arch/arm/mach-orion5x/Kconfig b/arch/arm/mach-orion5x/Kconfig
index 117cb87..6b94cdc 100644
--- a/arch/arm/mach-orion5x/Kconfig
+++ b/arch/arm/mach-orion5x/Kconfig
@@ -65,6 +65,44 @@ config MACH_LINKSTATION_MINI
 	  Say 'Y' here if you want your kernel to support the
 	  Buffalo Linkstation Mini platform.
 
+config MACH_LINKSTATION_PRODUO
+	bool "Buffalo Linkstation Pro Duo"
+	select I2C_BOARDINFO
+	help
+	  Say 'Y' here if you want your kernel to support the
+	  Buffalo Linkstation Pro Duo platform.
+
+	  LS-W1.0TGL/R1 is the general model number. There
+	  is no /R3 models, as /R1 stands for RAID1.
+	  There are two hardware revisions of the product.
+
+	  The first revision has version 1.xx firmware, 64 MB RAM,
+	  a single USB port, a power BUTTON, an Auto/Manual
+	  power MODE SWITCH, and a RESET button.
+
+	  The second revision has version 3.xx firmware, 128 MB RAM,
+	  two USB ports, an Off/On/Auto power SWITCH, and a FUNCTION button.
+
+	choice
+		prompt "HW model"
+		depends on MACH_LINKSTATION_PRODUO
+		default MACH_LINKSTATION_PRODUO_REV1
+		default MACH_LINKSTATION_PRODUO_REV2
+
+		config MACH_LINKSTATION_PRODUO_REV1
+		    bool "Revision 1"
+		    help
+		      The first revision has version 1.xx firmware, 64 MB RAM,
+		      a single USB port, a power BUTTON, an Auto/Manual
+		      power MODE SWITCH, and a RESET button.
+
+		config MACH_LINKSTATION_PRODUO_REV2
+		    bool "Revision 2"
+		    help
+		      The second revision has version 3.xx firmware, 128 MB RAM,
+		      two USB ports, an Off/On/Auto power SWITCH, and a FUNCTION button.
+	endchoice
+
 config MACH_TS409
 	bool "QNAP TS-409"
 	help
diff --git a/arch/arm/mach-orion5x/Makefile b/arch/arm/mach-orion5x/Makefile
index cc3f964..c76f950 100644
--- a/arch/arm/mach-orion5x/Makefile
+++ b/arch/arm/mach-orion5x/Makefile
@@ -5,6 +5,7 @@ obj-$(CONFIG_MACH_KUROBOX_PRO)	+= kurobox_pro-setup.o
 obj-$(CONFIG_MACH_TERASTATION_PRO2)	+= terastation_pro2-setup.o
 obj-$(CONFIG_MACH_LINKSTATION_PRO) += kurobox_pro-setup.o
 obj-$(CONFIG_MACH_LINKSTATION_MINI) += lsmini-setup.o
+obj-$(CONFIG_MACH_LINKSTATION_PRODUO) += lsproduo-setup.o
 obj-$(CONFIG_MACH_DNS323)	+= dns323-setup.o
 obj-$(CONFIG_MACH_TS209)	+= ts209-setup.o tsx09-common.o
 obj-$(CONFIG_MACH_TS409)	+= ts409-setup.o tsx09-common.o
